src/errors/validation_exception.rs

```diff
@@ -28,23 +29,45 @@ use super::ValError;
 #[derive(Clone)]
 #[cfg_attr(debug_assertions, derive(Debug))]
 pub struct ValidationError {
-    line_errors: Vec<PyLineError>,
+    line_errors: Vec<Arc<PyLineError>>,
+    validation_errors: Vec<Arc<ValidationError>>,
     error_mode: ErrorMode,
-    title: PyObject,
+    #[pyo3(get)]
+    message: PyObject,
+    #[pyo3(get)]
+    loc_prefix: Location,
 }
 
 impl ValidationError {
-    pub fn new(line_errors: Vec<PyLineError>, title: PyObject, error_mode: ErrorMode) -> Self {
+    pub fn new(line_errors: Vec<PyLineError>, message: PyObject, error_mode: ErrorMode) -> Self {
+        Self {
+            line_errors: line_errors.into_iter().map(Arc::new).collect(),
+            validation_errors: Vec::new(),
+            message,
+            error_mode,
+            loc_prefix: Location::Empty,
+        }
+    }
+
+    pub fn new_with_val_errors(
+        line_errors: Vec<Arc<PyLineError>>,
+        validation_errors: Vec<Arc<ValidationError>>,
+        message: PyObject,
+        error_mode: ErrorMode,
+        loc_prefix: Location,
+    ) -> Self {
```

Comment:
Adding a new constructor so that we can gradually migrate our internals instead of being forced to do it all at once.